fairscale/experimental/nn/mevo.py

```diff
@@ -0,0 +1,483 @@
+            grads[ctx.w_idx].add_(w.grad)
+        return i.grad, None, None, None, None, None
+
+
+class GetSumFunction(torch.autograd.Function):
+    """Custom checkpointed function to get sum-per-token from an input and a weight."""
+
+    @staticmethod
+    def get_sum(i: torch.Tensor, w: torch.Tensor, maxs: torch.Tensor, fp: bool) -> torch.Tensor:
+        """ i: (split-of-tokens, d_model), w: (split-of-vocabs, d_model)"""
+        _s = torch.matmul(i, w.T)
+        if fp:
+            _s = _s.float()
+        _s = (_s - maxs.reshape(-1, 1)).exp().sum(dim=1)
+        return _s
+
+    @staticmethod
+    def forward(  # type: ignore
+        ctx: Any,
+        i: torch.Tensor,
+        w: torch.Tensor,
+        maxs: torch.Tensor,
+        kernel_obj: "MemoryEfficientVocabOutput",
+        w_idx: int,
+        w_split_size: int,
+        split_dim: int,
+    ) -> torch.Tensor:
+        """Forward function that computes the sum, without saving activations."""
+        if DEBUG and dist.is_initialized() and dist.get_rank() == 0:
+            print("DEBUG sum fwd")
+        ctx.save_for_backward(i, w, maxs)
+        ctx.kernel_obj = kernel_obj
+        ctx.w_idx = w_idx
+        ctx.w_split_size = w_split_size
+        assert split_dim == 0
+        with torch.no_grad():
+            return GetSumFunction.get_sum(i, w, maxs, kernel_obj.fp_sum)
+
+    @staticmethod
+    def backward(ctx: Any, *args: Any) -> Any:
+        """Recompute the forward sum and backward grad.
+
+           Accumulate the grad to the right split of the full grad.
+        """
+        if DEBUG and dist.is_initialized() and dist.get_rank() == 0:
+            print("DEBUG sum bwd")
+        assert len(args) == 1
+        # Gradients should already exist due to TargetScoreFunction's backward.
+        assert ctx.kernel_obj.proj_weight.grad is not None
+
+        # Get saved i, w, and maxs.
+        i, w, maxs = ctx.saved_tensors
+        assert i.requires_grad
+        assert w.requires_grad
+        assert maxs.requires_grad
+        i = i.detach().requires_grad_(True)
+        w = w.detach().requires_grad_(True)
+        maxs = maxs.detach().requires_grad_(True)
+
+        # Forward + backward again.
+        with torch.enable_grad():
+            sums = GetSumFunction.get_sum(i, w, maxs, ctx.kernel_obj.fp_sum)
+        torch.autograd.backward(sums, *args)
+
+        # Accumulate the grads.
+        assert w.grad is not None
+        with torch.no_grad():
+            grads = torch.split(ctx.kernel_obj.proj_weight.grad, ctx.w_split_size)
+            grads[ctx.w_idx].add_(w.grad)
+        return i.grad, None, maxs.grad, None, None, None, None
+
+
+class TargetScoreFunction(torch.autograd.Function):
+    """Custom checkpointed function to compute the target score."""
+
+    @staticmethod
+    def get_target_score(i: torch.Tensor, w: torch.Tensor, target: torch.Tensor, fp: bool) -> torch.Tensor:
```

Comment:
can we rename i and w across this file to something that it represents? it might be easier to grok this code.

Reply:
Alright, I added an explanation of them at the first occurrence.